From 688b62fb3fa84d6446a74c13c93ddd33d87f5fe3 Mon Sep 17 00:00:00 2001
From: monsta <monsta@inbox.ru>
Date: Thu, 19 Feb 2015 14:11:39 +0300
Subject: [PATCH] daemon: fixed logic in a couple of places

---
 src/daemon/daemon.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/daemon/daemon.c b/src/daemon/daemon.c
index 49368f2..239bd02 100644
--- a/src/daemon/daemon.c
+++ b/src/daemon/daemon.c
@@ -342,7 +342,7 @@ static void on_popup_location_changed(GSettings *settings, gchar *key, NotifyDae
 	{
 		int j;
 
-		for (j = 0; j < daemon->priv->n_screens; j++)
+		for (j = 0; j < daemon->priv->screens[i]->n_stacks; j++)
 		{
 			NotifyStack* stack;
 			stack = daemon->priv->screens[i]->stacks[j];
@@ -399,7 +399,7 @@ static void destroy_screens(NotifyDaemon* daemon)
 		gdkwindow = gdk_screen_get_root_window(screen);
 		gdk_window_remove_filter(gdkwindow, (GdkFilterFunc) screen_xevent_filter, daemon->priv->screens[i]);
 
-		for (j = 0; i < daemon->priv->screens[i]->n_stacks; j++)
+		for (j = 0; j < daemon->priv->screens[i]->n_stacks; j++)
 		{
 			notify_stack_destroy(daemon->priv->screens[i]->stacks[j]);
 			daemon->priv->screens[i]->stacks[j] = NULL;

